lib/web_ui/lib/src/engine/embedder.dart

```diff
@@ -180,102 +133,32 @@ class FlutterViewEmbedder {
-    setElementStyle(bodyElement, 'color', 'red');
-
+    _applicationDom.setHostAttribute('flt-build-mode', buildMode);
+    _applicationDom.setHostAttribute('flt-application-dom', _applicationDom.type);
     // TODO(mdebbar): Disable spellcheck until changes in the framework and
     // engine are complete.
-    bodyElement.spellcheck = false;
-
-    for (final DomElement viewportMeta
-        in domDocument.head!.querySelectorAll('meta[name="viewport"]')) {
-      if (assertionsEnabled) {
-        // Filter out the meta tag that the engine placed on the page. This is
-        // to avoid UI flicker during hot restart. Hot restart will clean up the
-        // old meta tag synchronously with the first post-restart frame.
-        if (!viewportMeta.hasAttribute('flt-viewport')) {
-          print(
-            'WARNING: found an existing <meta name="viewport"> tag. Flutter '
-            'Web uses its own viewport configuration for better compatibility '
-            'with Flutter. This tag will be replaced.',
-          );
-        }
-      }
-      viewportMeta.remove();
-    }
+    _applicationDom.setHostAttribute('spellcheck', 'false');
+
+    // Set the global styles needed by flutter.
+    _applicationDom.setHostStyles(font: defaultCssFont);
+
+    // Set meta-viewport
+    _applicationDom.applyViewportMeta();
 
-    // This removes a previously created meta tag. Note, however, that this does
-    // not remove the meta tag during hot restart. Hot restart resets all static
-    // variables, so this will be null upon hot restart. Instead, this tag is
-    // removed by _clearOnHotRestart.
-    _viewportMeta?.remove();
-    _viewportMeta = createDomHTMLMetaElement()
-      ..setAttribute('flt-viewport', '')
-      ..name = 'viewport'
-      ..content = 'width=device-width, initial-scale=1.0, '
-          'maximum-scale=1.0, user-scalable=no';
-    domDocument.head!.append(_viewportMeta!);
-
-    // IMPORTANT: the glass pane element must come after the scene element in the DOM node list so
-    //            it can intercept input events.
-    _glassPaneElement?.remove();
-    final DomElement glassPaneElement = domDocument.createElement(_glassPaneTagName);
+    // Create and inject the [_glassPaneElement].
+    final DomElement glassPaneElement = domDocument.createElement(glassPaneTagName);
     _glassPaneElement = glassPaneElement;
-    glassPaneElement.style
-      ..position = 'absolute'
-      ..top = '0'
-      ..right = '0'
-      ..bottom = '0'
-      ..left = '0';
 
-    // This must be appended to the body, so the engine can create a host node
-    // properly.
-    bodyElement.append(glassPaneElement);
+    // This must be appended to the applicationDom now, so the engine can create
+    // a host node (ShadowDOM or a fallback) next.
+    //
+    // The applicationDom will take care of cleaning up the glassPane on hot
+    // restart.
+    _applicationDom.attachGlassPane(glassPaneElement);
```

Comment:
Here too, I think `ApplicationDom` can fully own the glass pane element, instead of having something from outside create and set it. Similarly, it should own the initialization of `PointerBinding` and metrics change listener.

I'm OK with making this transition gradually, but I think the final design should consolidate ownership of view resources into the "application DOM" abstraction. Splitting it between the view and the embedder will make things more complex.

Reply:
For now, this is the line that I wanted to keep:

* ApplicationDOM knows *where* the GlassPane ends up in the page. This needs to know the hostElement and behaves differently.
* ViewEmbedder knows *what* the GlassPane made of, and how to make it. This doesn't care about the hostElement.

If your idea is to keep moving logic from the ViewEmbedder to the ApplicationDOM, shouldn't we just create a `FullPageViewEmbedder` and a `CustomElementViewEmbedder` and take the plunge now?

Reply:
Ah, I see where you are going with this. I guess in my mind `ViewEmbedder` is still a global singleton, which manages multiple views. In contrast, `ApplicationDom` is not a singleton and is agnostic of embedding strategy (full-page, custom element, multiple custom elements). So all my suggestions revolve around moving any view-specific initialization logic from single scope to per-view scope.

If the idea is that `ViewEmbedder` ceases to be singleton, then it might scale. However:

> shouldn't we just create a FullPageViewEmbedder and a CustomElementViewEmbedder and take the plunge now?

If this is not boiling too many oceans, then it looks like the cleanest solutions. I still think we need a root-level object for managing all the views to avoid top-level statics. To me `ViewEmbedder` looks like a good candidate for that, with `EngineView`, or some such, being "managed objects".

Reply:
"If this is not boiling too many oceans" -> I took a look at the code, and this is *definitely* boiling too many oceans.

Form what I've seen (and I might be wrong, because I've only read code and not really *traced* most of it running:

#### Rendering:

* The EnginePlatformDispatcher knows how to handle [multiple views](https://github.com/flutter/engine/blob/7e03ebcbee0dfb4d361618deeeb3d4595156ca65/lib/web_ui/lib/src/engine/window.dart#L48-L51) and their configurations; however that is hardcoded to a single `engine.window`. Also, the view configuration is never really set to anything other than the default values.
    * Views have access to the PlatformDispatcher. The PlatformDispatcher can [`render` a `Scene`](https://github.com/flutter/engine/blob/7e03ebcbee0dfb4d361618deeeb3d4595156ca65/lib/web_ui/lib/src/engine/platform_dispatcher.dart#L661-L664) into its `View` through a global `renderer` object, but we never use the incoming FlutterView to render the scene.
        * The renderer knows about the global `embedder` and its [`addSceneToSceneHost` method](https://github.com/flutter/engine/blob/7e03ebcbee0dfb4d361618deeeb3d4595156ca65/lib/web_ui/lib/src/engine/canvaskit/renderer.dart#L75) that glues this all together, however this bit is definitely not multi-view compliant, because the embedder controls the embedding strategy.
* The Views have a "peculiar" inheritance. For all intents and purposes, on the web, we only care about the `EngineFlutterWindow` class (there's a `EngineSingletonFlutterWindow` that exposes the DPI of the device, but that should probably end up disappearing)
    * The default `engine.view` (of type `EngineSingletonFlutterWindow`) is returned from`ui.view`.

If there's interest in the multi-view project, I think I'd start by fixing how the renderers control the targets of their scenes. Instead, the renderer should paint through `view.viewConfiguration.embeddingStrategy.renderScene(scene)` (or `getEmbeddingStrategy(view).renderScene(scene);`).

> I think the final design should consolidate ownership of view resources into the "application DOM" abstraction. Splitting it between the view and the embedder will make things more complex.

I think the centralization of things needs to happen in the web specializations of the "View" classes, that already exist (the URL Strategy is already there; the Embedding Strategy should probably live inside each View too).

The current EmbeddingStrategy objects can be maintained as they are (they're small and pretty much single-purpose), but you're right that the global `view_embedder` has its days numbered :P

PS: It **does** makes sense in the above world that each EmbeddingStrategy owns its own glassPane element. I'll add it there, and for now, have the Singleton Embedder consult its Singleton EmbeddingStrategy for the glassPaneElement to maintain its API.

Takeaway: I'm renaming `ApplicationDom` to `EmbeddingStrategy` :)

Reply:
I backpedaled on the EmbeddingStrategy owning the glassPane element. I think there's still value in having a standard way of creating a Flutter App regardless of where it is injected. For now I left it in the View Embedder, but we should have a class that "knows how to create and assemble a flutter app" in a unified way.

Reply:
Here's the future: https://github.com/flutter/flutter/issues/117098